keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1387 @@
+CSI API responses.
+
+Enhancements for `NodePublishVolumeRequest`: new field `runtime_supports_mount`
+```
+message NodePublishVolumeRequest {
+  // The ID of the volume to publish. This field is REQUIRED.
+  string volume_id = 1;
+  ...
+  // Indicates Container Runtime supports mounting a file system
+  // This field is OPTIONAL.
+  bool runtime_supports_mount = 6 [(alpha_field) = true];
+}
+```
+
+Enhancements for `NodeGetVolumeStatsRequest`: new field `runtime_supports_stats`
+```
+message NodeGetVolumeStatsRequest {
+  // The ID of the volume. This field is REQUIRED.
+  string volume_id = 1;
+  ...
+  // Indicates Container Runtime supports reporting stats and
+  // condition of the file system on the volume
+  // This field is OPTIONAL.
+  bool runtime_supports_stats = 4 [(alpha_field) = true];
+```
+
+Enhancements for `NodeExpandVolumeRequest`: new field `runtime_supports_expand`
+```
+message NodeExpandVolumeRequest {
+  // The ID of the volume to publish. This field is REQUIRED.
+  string volume_id = 1;
+  ...
+  // Indicates Container Runtime supports expanding the file system
+  // on the volume
+  // This field is OPTIONAL.
+  bool runtime_supports_expand = 7 [(alpha_field) = true];
+}
+```
+
+#### Enhancements in CSI API Responses
+
+Corresponding to the above enhancements in CSI API Requests, new fields are
+necessary in the CSI node API responses: `NodePublishVolumeResponse`,
+`NodeGetVolumeStatsResponse` and `NodeVolumeExpandResponse`. The CSI plugin needs to
+indicate to the Kubelet (or another Container Orchestrator from the orchestrator
+agnostic CSI perspective) that the runtime (associated with the pod mounting the
+PV) should be involved in processing of the API along with relevant parameters
+to identify the volume and process the API call.
+
+Enhancements for `NodePublishVolumeResponse`: new optional field `runtime_mount_info`
+specifying details of how and what to mount on a block device or network share.
+```
+message FileSystemMountInfo {
+  // Source device or network share (e.g. /dev/sdX, srv:/export) whose mount
+  // is deferred to a container runtime that is capable of executing the mount.
+  // This field is REQUIRED
+  string source = 1 [(alpha_field) = true];
+  // Type of the filesystem to mount (e.g. xfs, ext4, nfs, ntfs) on the specified
+  // source.
+  // This field is REQUIRED
+  string type = 2 [(alpha_field) = true];
+  // Mount options supported by the filesystem to be used for the specified source.
+  // This field is OPTIONAL.
+  map<string, string> options = 3 [(alpha_field) = true];
+}
+
+message NodePublishVolumeResponse {
+  // Specifies details of how to mount a file system on a source device or
+  // network share when SP defers file system mounts to a container runtime.
+  // A SP MUST populate this if runtime_supports_mount was set in
+  // NodePublishVolumeRequest and SP is capable of deferring filesystem mount to
+  // the container runtime.
```

Comment:
A new capability needs to be defined. Otherwise kubelet won't know whether the SP checks `runtime_supports_mount` and sets this field.